chain/client/src/client.rs

```diff
@@ -2374,12 +2374,11 @@ impl Client {
                 state_sync_info.shards.iter().map(|tuple| tuple.0).collect();
             // Notify each shard to sync.
             if notify_state_sync {
-                let shard_layout = self
-                    .epoch_manager
-                    .get_shard_layout(&epoch_id)
-                    .expect("Cannot get shard layout");
-                for &shard_id in &tracking_shards {
-                    let shard_uid = ShardUId::from_shard_id_and_layout(shard_id, &shard_layout);
+                for shard_uid in
+                    self.epoch_manager.shard_ids_to_uids(&tracking_shards, &epoch_id).unwrap()
+                {
+                    // Explicitly unload mem-trie for shard that we start state sync, even if the shard was tracked in the same epoch.
+                    self.runtime_adapter.unload_mem_trie(&shard_uid);
```

Comment:
The `notify_state_sync` flag is set only when the node first starts to process state sync for a new epoch (or `sync_hash`). 
The `tracking_shards` contains all the shards that need to be tracked in the next epoch. Some of those shards are already being tracked and may be up to date with the chain head, maybe even used to validate in the current epoch. 

Is it your intention that the node will drop those shards as well?

Reply:
Resharding is implemented as a state sync in some sense and it may lead to having a state sync for a shard that is already tracked. This admittedly could have been done in a cleaner way. Anyway I wouldn't worry about it as state sync and resharding are not integrated anyway.

Reply:
I moved `unload_mem_trie()` just before flat state is cleared. Now flat state and memtrie are loaded / unloaded together, and we do not unload mem-trie too early or when we should not.